python/src/graphql.rs

```diff
@@ -1,87 +1,590 @@
+                        tokio_sender
+                            .blocking_send(())
+                            .expect("Failed to send cancellation signal");
+                    });
+                    handler.wait().await
+                })
+        });
+
+        Ok(PyRunningRaphtoryServer::new(join_handle, sender, port))
+    }
+
+    /// Run the server on the default port until completion.
+    pub fn run(slf: PyRefMut<Self>) -> PyResult<()> {
+        wait_server(&mut Self::start(slf)?.0)
+    }
+
+    /// Run the server on the port `port` until completion.
+    pub fn run_with_port(slf: PyRefMut<Self>, port: u16) -> PyResult<()> {
+        wait_server(&mut Self::start_with_port(slf, port)?.0)
+    }
+}
+
+fn take_sever_ownership(mut server: PyRefMut<PyRaphtoryServer>) -> PyResult<RaphtoryServer> {
+    let new_server = server.0.take().ok_or_else(|| {
+        PyException::new_err(
+            "Server object has already been used, please create another one from scratch",
+        )
+    })?;
+    Ok(new_server)
+}
+
+fn wait_server(running_server: &mut Option<ServerHandler>) -> PyResult<()> {
+    let owned_running_server = running_server
+        .take()
+        .ok_or_else(|| PyException::new_err(RUNNING_SERVER_CONSUMED_MSG))?;
+    owned_running_server
+        .join_handle
+        .join()
+        .expect("error when waiting for the server thread to complete")
+        .map_err(|e| adapt_err_value(&e))
+}
+
+const RUNNING_SERVER_CONSUMED_MSG: &str =
+    "Running server object has already been used, please create another one from scratch";
+
+/// A Raphtory server handler that also enables querying the server
+#[pyclass(name = "RunningRaphtoryServer")]
+pub(crate) struct PyRunningRaphtoryServer(Option<ServerHandler>);
+
+struct ServerHandler {
+    join_handle: JoinHandle<IoResult<()>>,
+    sender: Sender<()>,
+    client: PyRaphtoryClient,
+}
+
+impl PyRunningRaphtoryServer {
+    fn new(join_handle: JoinHandle<IoResult<()>>, sender: Sender<()>, port: u16) -> Self {
+        let url = format!("http://localhost:{port}");
+        Self(Some(ServerHandler {
+            join_handle,
+            sender,
+            client: PyRaphtoryClient::new(url),
+        }))
+    }
+
+    fn apply_if_alive<O, F>(&self, function: F) -> PyResult<O>
+    where
+        F: FnOnce(&ServerHandler) -> PyResult<O>,
+    {
+        match &self.0 {
+            Some(handler) => function(handler),
+            None => Err(PyException::new_err(RUNNING_SERVER_CONSUMED_MSG)),
+        }
+    }
+}
+
+#[pymethods]
+impl PyRunningRaphtoryServer {
+    /// Stop the server.
+    pub(crate) fn stop(&self) -> PyResult<()> {
+        self.apply_if_alive(|handler| {
+            handler
+                .sender
+                .send(())
+                .expect("Failed when sending cancellation signal");
+            Ok(())
+        })
+    }
+
+    /// Wait until server completion.
+    pub(crate) fn wait(mut slf: PyRefMut<Self>) -> PyResult<()> {
```

Comment:
Does this block forever? I.e. can I cancel it?

Reply:
if there is a Ctrl+C event, that will cause the server to stop and then this function will finish

Reply:
anyways, the main utility of this function is when you call `server.stop()`, which returns immediately, and you want to make sure that it is completely stopped before going on with the program

Reply:
I'm just checking and it seems it is not working as I expected, will fix it